relay-config/src/redis.rs

```diff
@@ -128,12 +130,90 @@ impl From<RedisConfigFromFile> for RedisConfig {
     }
 }
 
-impl<'de> Deserialize<'de> for RedisConfig {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
-    where
-        D: serde::Deserializer<'de>,
-    {
-        RedisConfigFromFile::deserialize(deserializer).map(Into::into)
+/// Configurations for the various Redis pools used by Relay.
+#[derive(Clone, Debug, Deserialize, Serialize, Eq, PartialEq)]
+#[serde(untagged)]
+pub enum RedisConfigs {
+    /// All pools should be configured the same way.
+    Unified(RedisConfig),
+    /// Individual configurations for each pool.
+    Individual {
+        /// Configuration for the `project_configs` pool.
+        project_configs: Box<RedisConfig>,
+        /// Configuration for the `cardinality` pool.
+        cardinality: Box<RedisConfig>,
+        /// Configuration for the `quotas` pool.
+        quotas: Box<RedisConfig>,
+        /// Configuration for the `misc` pool.
+        misc: Box<RedisConfig>,
+    },
+}
+
+/// Helper struct bundling connections and options for the various Redis pools.
+#[derive(Clone, Debug)]
+pub struct RedisPoolConfigs<'a> {
+    /// Configuration for the `project_configs` pool.
+    pub project_configs: (&'a RedisConnection, RedisConfigOptions),
+    /// Configuration for the `cardinality` pool.
+    pub cardinality: (&'a RedisConnection, RedisConfigOptions),
+    /// Configuration for the `quotas` pool.
+    pub quotas: (&'a RedisConnection, RedisConfigOptions),
+    /// Configuration for the `misc` pool.
+    pub misc: (&'a RedisConnection, RedisConfigOptions),
+}
+
+pub(super) fn create_redis_pool(
+    config: &RedisConfig,
+    cpu_concurrency: usize,
+) -> (&RedisConnection, RedisConfigOptions) {
+    let options = RedisConfigOptions {
+        max_connections: config
+            .options
+            .max_connections
+            .unwrap_or(cpu_concurrency as u32 * 2)
+            .min(crate::redis::DEFAULT_MIN_MAX_CONNECTIONS),
```

Comment:
```suggestion
pub(super) fn create_redis_pool(
    config: &RedisConfig,
    default_connections: usize,
) -> (&RedisConnection, RedisConfigOptions) {
    let options = RedisConfigOptions {
        max_connections: config
            .options
            .max_connections
            .unwrap_or(default_connections)
```

If the function takes `default_connections` and we move the logic out, we can have more easily different defaults for per pool.